File: android/app/src/main/java/com/roxfollow/app/ads/UnityAdsManager.java
package com.roxfollow.app.ads;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import com.unity3d.ads.IUnityAdsInitializationListener;
import com.unity3d.ads.IUnityAdsLoadListener;
import com.unity3d.ads.IUnityAdsShowListener;
import com.unity3d.ads.UnityAds;
import com.unity3d.ads.UnityAdsShowOptions;

public class UnityAdsManager {

    private static final String TAG = "UnityAdsManager";

    private static final String GAME_ID = "800368206";
    public static final String REWARDED_PLACEMENT_ID = "Rewarded_Android";

    private static final boolean TEST_MODE = false;

    private static boolean isInitialized = false;
    private static volatile boolean rewardedAdLoaded = false;

    public interface RewardCallback {
        void onReward();
    }

    public interface FailCallback {
        void onFailedOrSkipped();
    }

    public static void initialize(Context context) {
        if (isInitialized) return;
        UnityAds.initialize(context, GAME_ID, TEST_MODE, new IUnityAdsInitializationListener() {
            @Override
            public void onInitializationComplete() {
                Log.d(TAG, "Unity Ads initialized successfully");
                isInitialized = true;
                loadRewardedAd();
            }

            @Override
            public void onInitializationFailed(UnityAds.UnityAdsInitializationError error, String message) {
                Log.e(TAG, "Unity Ads init failed: " + error + " - " + message);
            }
        });
    }

    public static void loadRewardedAd() {
        rewardedAdLoaded = false;
        UnityAds.load(REWARDED_PLACEMENT_ID, new IUnityAdsLoadListener() {
            @Override
            public void onUnityAdsAdLoaded(String placementId) {
                Log.d(TAG, "Rewarded ad loaded: " + placementId);
                rewardedAdLoaded = true;
            }

            @Override
            public void onUnityAdsFailedToLoad(String placementId, UnityAds.UnityAdsLoadError error, String message) {
                Log.e(TAG, "Rewarded ad failed to load: " + error + " - " + message);
                rewardedAdLoaded = false;
            }
        });
    }

    public static boolean isRewardedAdReady() {
        return rewardedAdLoaded;
    }

    public static void showRewardedAd(Activity activity, RewardCallback onReward, FailCallback onFailedOrSkipped) {
        rewardedAdLoaded = false;
        UnityAds.show(activity, REWARDED_PLACEMENT_ID, new UnityAdsShowOptions(), new IUnityAdsShowListener() {
            @Override
            public void onUnityAdsShowFailure(String placementId, UnityAds.UnityAdsShowError error, String message) {
                Log.e(TAG, "Rewarded ad show failed: " + error + " - " + message);
                if (onFailedOrSkipped != null) onFailedOrSkipped.onFailedOrSkipped();
                loadRewardedAd();
            }

            @Override
            public void onUnityAdsShowStart(String placementId) {}

            @Override
            public void onUnityAdsShowClick(String placementId) {}

            @Override
            public void onUnityAdsShowComplete(String placementId, UnityAds.UnityAdsShowCompletionState state) {
                if (state == UnityAds.UnityAdsShowCompletionState.COMPLETED) {
                    if (onReward != null) onReward.onReward();
                } else {
                    if (onFailedOrSkipped != null) onFailedOrSkipped.onFailedOrSkipped();
                }
                loadRewardedAd();
            }
        });
    }
}

File: app/src/main/java/com/example/ads/UnityAdsManager.kt
package com.example.ads

import android.app.Activity
import android.content.Context
import android.util.Log
import com.unity3d.ads.IUnityAdsInitializationListener
import com.unity3d.ads.IUnityAdsLoadListener
import com.unity3d.ads.IUnityAdsShowListener
import com.unity3d.ads.UnityAds
import com.unity3d.ads.UnityAdsShowOptions

/**
 * Central place for Unity Ads setup — Rox Follow app.
 * Game ID and Rewarded placement are the real, live values from the Unity dashboard.
 * Only rewarded video is used (no interstitial, no banner) by design.
 */
object UnityAdsManager {

    private const val TAG = "UnityAdsManager"

    private const val GAME_ID = "800368206"
    const val REWARDED_PLACEMENT_ID = "Rewarded_Android"

    private const val TEST_MODE = false

    private var isInitialized = false

    fun initialize(context: Context) {
        if (isInitialized) return
        UnityAds.initialize(
            context,
            GAME_ID,
            TEST_MODE,
            object : IUnityAdsInitializationListener {
                override fun onInitializationComplete() {
                    Log.d(TAG, "Unity Ads initialized successfully")
                    isInitialized = true
                    loadRewardedAd()
                }

                override fun onInitializationFailed(
                    error: UnityAds.UnityAdsInitializationError?,
                    message: String?
                ) {
                    Log.e(TAG, "Unity Ads init failed: $error - $message")
                }
            }
        )
    }

    // ---------- Rewarded ----------

    @Volatile
    private var rewardedAdLoaded = false

    fun loadRewardedAd() {
        rewardedAdLoaded = false
        UnityAds.load(REWARDED_PLACEMENT_ID, object : IUnityAdsLoadListener {
            override fun onUnityAdsAdLoaded(placementId: String?) {
                Log.d(TAG, "Rewarded ad loaded: $placementId")
                rewardedAdLoaded = true
            }

            override fun onUnityAdsFailedToLoad(
                placementId: String?,
                error: UnityAds.UnityAdsLoadError?,
                message: String?
            ) {
                Log.e(TAG, "Rewarded ad failed to load: $error - $message")
                rewardedAdLoaded = false
            }
        })
    }

    fun showRewardedAd(
        activity: Activity,
        onReward: () -> Unit,
        onFailedOrSkipped: (() -> Unit)? = null
    ) {
        rewardedAdLoaded = false
        UnityAds.show(
            activity,
            REWARDED_PLACEMENT_ID,
            UnityAdsShowOptions(),
            object : IUnityAdsShowListener {
                override fun onUnityAdsShowFailure(
                    placementId: String?,
                    error: UnityAds.UnityAdsShowError?,
                    message: String?
                ) {
                    Log.e(TAG, "Rewarded ad show failed: $error - $message")
                    onFailedOrSkipped?.invoke()
                    loadRewardedAd()
                }

                override fun onUnityAdsShowStart(placementId: String?) {}

                override fun onUnityAdsShowClick(placementId: String?) {}

                override fun onUnityAdsShowComplete(
                    placementId: String?,
                    state: com.unity3d.ads.UnityAds.UnityAdsShowCompletionState?
                ) {
                    if (state == com.unity3d.ads.UnityAds.UnityAdsShowCompletionState.COMPLETED) {
                        onReward()
                    } else {
                        onFailedOrSkipped?.invoke()
                    }
                    loadRewardedAd()
                }
            }
        )
    }

    fun isRewardedAdReady(): Boolean = rewardedAdLoaded
}
